extlib/millhone/src/cmd/analyze.rs

```diff
@@ -36,123 +38,157 @@ pub struct Subcommand {
-    let walk = WalkDir::new(root)
-        // Follow symlinks; loops are yielded as errors automatically.
-        .follow_links(true)
-        // Not chosen for a specific reason, just seems reasonable.
-        .max_depth(1000)
-        // Just make the walk deterministic (per directory anyway).
-        .sort_by_file_name();
-
-    let mut total_count_entries = 0usize;
-    let mut total_count_snippets = 0usize;
-    let mut total_count_files = 0usize;
-
     if opts.output().exists() {
         if opts.overwrite_output {
             std::fs::remove_dir_all(opts.output()).context("remove existing output directory")?;
+            debug!("removed existing output dir: {}", opts.output().display());
         } else {
             bail!(
-                "The output directory '{}' already exists.",
+                "the output directory '{}' already exists.",
                 opts.output().display(),
             );
         }
     }
     if !opts.output().exists() {
         std::fs::create_dir_all(opts.output()).context("create output directory")?;
+        debug!("created output dir: {}", opts.output().display());
     }
 
-    // Future enhancement: walk and analyze in parallel with rayon.
+    let creds = opts.auth.as_credentials();
+    let client = ApiClientV1::authenticated(endpoint, creds);
+    let root = opts.extract().target();
     let snippet_opts = opts.extract().into();
-    for entry in walk.into_iter() {
-        total_count_entries += 1;
-        let Some(entry) = super::unwrap_dir_entry(entry) else {
-            continue;
-        };
-
-        let path = if entry.path_is_symlink() {
-            let path = entry.path();
-            fs::read_link(path)
-                .wrap_err_with(|| format!("resolve symlink of '{}'", path.display()))?
-        } else {
-            entry.path().to_path_buf()
-        };
-
-        if !path.is_file() {
-            debug!(path = %path.display(), "skipped: not a file");
-            continue;
-        }
-
-        total_count_files += 1;
-        info!(path = %path.display(), "analyze");
 
-        let snippets = ContentSnippet::from_file(root, &snippet_opts, &path)
-            .wrap_err_with(|| format!("process '{}'", path.display()))?;
+    let total_count_entries = AtomicCounter::default();
+    let total_count_snippets = AtomicCounter::default();
+    let total_count_files = AtomicCounter::default();
 
-        info!(snippet_count = %snippets.len(), "extracted snippets");
-        total_count_snippets += snippets.len();
-        if snippets.is_empty() {
-            continue;
-        }
+    WalkDir::new(root)
+        // Follow symlinks; loops are yielded as errors automatically.
+        .follow_links(true)
+        // Not chosen for a specific reason, just seems reasonable.
+        .max_depth(1000)
+        // Just make the walk deterministic (per directory anyway).
+        .sort_by_file_name()
+        .into_iter()
+        .inspect(|_| total_count_entries.increment())
```

Comment:
Is there a difference between `inspect` and `tap`? It looks to me like they do the same thing, but `inspect` is a standard method on iterators, so it's idiomatic to use them there. Is that right?

Reply:
`inspect` is an iterator adaptor (so for a given `Iterator<Item = T>`, its closure receives `&T`), whereas `tap` takes the value (so for a given `T`, its closure receives `&T`).

This looks really similar but the effect on iterators is such that for e.g. `vec![1].into_iter().tap(|x| {})`, `x` is actually `&impl Iterator<Item = usize>`, _not_ what you're actually after (which is `&usize`, and accessible via `inspect`).

Reply:
We can also see this in the different syntax:

- `tap`: `fn tap(self, func: impl FnOnce(&Self)) -> Self`
- `inspect`: `fn inspect<F: FnMut(&Self::Item)>(self, f: F) -> Inspect<Self, F>` (also note `Inspect` implements `Iterator`).

Without going too much into it, from there we can see that `tap` actually takes ownership of `self` (which is whatever it's being invoked on). Consequently, if invoked on an iterator, it gets the whole iterator and not just one item.